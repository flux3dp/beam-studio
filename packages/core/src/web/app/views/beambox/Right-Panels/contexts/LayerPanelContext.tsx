import React, { createContext, useCallback, useEffect, useMemo, useState } from 'react';

import { pipe } from 'remeda';

import { promarkModels } from '@core/app/actions/beambox/constant';
import { LayerModule } from '@core/app/constants/layer-module/layer-modules';
import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import useWorkarea from '@core/helpers/hooks/useWorkarea';
import doLayersContainsVector from '@core/helpers/layer/check-vector';
import { getData } from '@core/helpers/layer/layer-config-helper';
import { getLayerElementByName } from '@core/helpers/layer/layer-helper';
import useForceUpdate from '@core/helpers/use-force-update';

interface ILayerPanelContext {
  forceUpdate: () => void;
  forceUpdateSelectedLayers: () => void;
  hasGradient: boolean;
  hasVector: boolean;
  selectedLayers: string[];
  setSelectedLayers: (selectedLayers: string[]) => void;
}

export const LayerPanelContext = createContext<ILayerPanelContext>({
  forceUpdate: () => {},
  forceUpdateSelectedLayers: () => {},
  hasGradient: false,
  hasVector: false,
  selectedLayers: [],
  setSelectedLayers: () => {},
});

const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

interface Props {
  children?: React.ReactNode;
}

export const LayerPanelContextProvider = ({ children }: Props): React.JSX.Element => {
  const [hasVector, setHasVector] = useState<boolean>(false);
  const [hasGradient, setHasGradient] = useState<boolean>(false);
  const [selectedLayers, setSelectedLayers] = useState<string[]>([]);
  const forceUpdate = useForceUpdate();
  const forceUpdateSelectedLayers = useCallback(() => setSelectedLayers([...selectedLayers]), [selectedLayers]);
  const workarea = useWorkarea();
  const isPromark = useMemo(() => promarkModels.has(workarea), [workarea]);
  const lazySetSelectedLayers = useCallback(
    (newLayers: string[]) => {
      if (newLayers.length === selectedLayers.length && newLayers.every((name, i) => name === selectedLayers[i])) {
        return;
      }

      const isUvExportable = newLayers.every(
        (layerName) => getData(getLayerElementByName(layerName), 'module') === LayerModule.UV_EXPORT,
      );

      layerPanelEventEmitter.emit('updateUvExportStatus', isUvExportable);

      setSelectedLayers(newLayers);
    },
    [selectedLayers, setSelectedLayers],
  );

  useEffect(() => {
    layerPanelEventEmitter.on('UPDATE_LAYER_PANEL', forceUpdate);

    return () => {
      layerPanelEventEmitter.removeListener('UPDATE_LAYER_PANEL', forceUpdate);
    };
    // eslint-disable-next-line hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    layerPanelEventEmitter.on('SET_SELECTED_LAYERS', lazySetSelectedLayers);

    return () => {
      layerPanelEventEmitter.removeListener('SET_SELECTED_LAYERS', lazySetSelectedLayers);
    };
  }, [lazySetSelectedLayers]);

  useEffect(() => {
    const getSelectedLayers = (response: { selectedLayers: string[] }) => {
      response.selectedLayers = selectedLayers;
    };

    layerPanelEventEmitter.on('GET_SELECTED_LAYERS', getSelectedLayers);

    const checkVector = () => {
      const newVal = doLayersContainsVector(selectedLayers);

      setHasVector(newVal);
    };

    layerPanelEventEmitter.on('CHECK_VECTOR', checkVector);

    return () => {
      layerPanelEventEmitter.off('GET_SELECTED_LAYERS', getSelectedLayers);
      layerPanelEventEmitter.off('CHECK_VECTOR', checkVector);
    };
  }, [selectedLayers]);

  useEffect(() => {
    const checkGradient = () => {
      if (isPromark) {
        const newVal = selectedLayers.some((layerName: string) =>
          pipe(
            //
            layerName,
            getLayerElementByName,
            (layer) => layer?.querySelector('image[data-shading="true"]'),
            Boolean,
          ),
        );

        setHasGradient(newVal);
      }
    };

    checkGradient();
    layerPanelEventEmitter.on('CHECK_GRADIENT', checkGradient);

    return () => {
      layerPanelEventEmitter.removeListener('CHECK_GRADIENT', checkGradient);
    };
  }, [selectedLayers, isPromark]);

  // move doLayersContainsVector and setVector in effect to avoid heavy render process
  // eslint-disable-next-line hooks/exhaustive-deps
  useEffect(() => {
    const newVal = doLayersContainsVector(selectedLayers);

    if (hasVector !== newVal) {
      setHasVector(newVal);
    }

    if (isPromark) {
      const newGradientVal = selectedLayers.some((layerName: string) =>
        pipe(
          //
          layerName,
          getLayerElementByName,
          (layer) => layer?.querySelector('image[data-shading="true"]'),
          Boolean,
        ),
      );

      if (newGradientVal !== hasGradient) {
        setHasGradient(newGradientVal);
      }
    }
  });

  return (
    <LayerPanelContext.Provider
      value={{
        forceUpdate,
        forceUpdateSelectedLayers,
        hasGradient,
        hasVector,
        selectedLayers,
        setSelectedLayers: lazySetSelectedLayers,
      }}
    >
      {children}
    </LayerPanelContext.Provider>
  );
};
