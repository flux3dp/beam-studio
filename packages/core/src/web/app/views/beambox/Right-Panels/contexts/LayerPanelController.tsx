import eventEmitterFactory from 'helpers/eventEmitterFactory';

const layerPanelEventEmitter = eventEmitterFactory.createEventEmitter('layer-panel');

const updateLayerPanel = (): void => {
  layerPanelEventEmitter.emit('UPDATE_LAYER_PANEL');
};

const getSelectedLayers = (): string[] => {
  const response = {
    selectedLayers: [],
  };
  layerPanelEventEmitter.emit('GET_SELECTED_LAYERS', response);
  return response.selectedLayers;
};

const setSelectedLayers = (selectedLayers: string[]): void => {
  layerPanelEventEmitter.emit('SET_SELECTED_LAYERS', selectedLayers);
};

const checkVector = (): void => {
  layerPanelEventEmitter.emit('CHECK_VECTOR');
};

const checkGradient = (): void => {
  layerPanelEventEmitter.emit('CHECK_GRADIENT');
};

export default {
  checkVector,
  checkGradient,
  updateLayerPanel,
  getSelectedLayers,
  setSelectedLayers,
};
