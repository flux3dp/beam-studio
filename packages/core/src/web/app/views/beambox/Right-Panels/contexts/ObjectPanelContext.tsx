/* eslint-disable reactRefresh/only-export-components */
import React, { useCallback, useEffect, useRef, useState } from 'react';

import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import useForceUpdate from '@core/helpers/use-force-update';
import type { DimensionKey, DimensionValues } from '@core/interfaces/ObjectPanel';

interface IObjectPanelContext {
  activeKey: null | string;
  dimensionValues: any;
  getDimensionValues: (
    response: {
      dimensionValues: any;
    },
    key?: DimensionKey,
  ) => void;
  polygonSides: number;
  updateActiveKey: (activeKey: null | string) => void;
  updateDimensionValues: (newValues: DimensionValues) => void;
  updateObjectPanel: () => void;
}

export const ObjectPanelContext = React.createContext<IObjectPanelContext>({
  activeKey: null,
  dimensionValues: {},
  getDimensionValues: () => {},
  polygonSides: 5,
  updateActiveKey: () => {},
  updateDimensionValues: () => {},
  updateObjectPanel: () => {},
});

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');
const minRenderInterval = 200;

interface Props {
  children?: React.ReactNode;
}

export const ObjectPanelContextProvider = ({ children }: Props): React.JSX.Element => {
  const forceUpdate = useForceUpdate();
  const [polygonSides, setPolygonSides] = useState(5);
  const [activeKey, setActiveKey] = useState<null | string>(null);
  const dimensionValues = useRef<DimensionValues>({});
  const lastUpdateTime = useRef(Date.now());
  const updateTimeout = useRef<NodeJS.Timeout | null>(null);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_POLYGON_SIDES', setPolygonSides);
    objectPanelEventEmitter.on('UPDATE_ACTIVE_KEY', setActiveKey);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_POLYGON_SIDES', setPolygonSides);
      objectPanelEventEmitter.removeListener('UPDATE_ACTIVE_KEY', setActiveKey);
    };
  }, []);

  const getDimensionValues = useCallback((response: { dimensionValues: any }, key?: DimensionKey) => {
    response.dimensionValues = key ? dimensionValues.current[key] : dimensionValues.current;
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('GET_DIMENSION_VALUES', getDimensionValues);

    return () => {
      objectPanelEventEmitter.removeListener('GET_DIMENSION_VALUES', getDimensionValues);
    };
  }, [getDimensionValues]);

  const updateDimensionValues = useCallback((newValues: DimensionValues) => {
    dimensionValues.current = {
      ...dimensionValues.current,
      ...newValues,
    };
  }, []);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_DIMENSION_VALUES', updateDimensionValues);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_DIMENSION_VALUES', updateDimensionValues);
    };
  }, [updateDimensionValues]);

  const getActiveKey = useCallback(
    (response: { activeKey: null | string }) => {
      response.activeKey = activeKey;
    },
    [activeKey],
  );

  useEffect(() => {
    objectPanelEventEmitter.on('GET_ACTIVE_KEY', getActiveKey);

    return () => {
      objectPanelEventEmitter.removeListener('GET_ACTIVE_KEY', getActiveKey);
    };
  }, [getActiveKey]);

  const updateObjectPanel = useCallback(() => {
    if (updateTimeout.current) clearTimeout(updateTimeout.current);

    const time = Date.now();

    if (time - lastUpdateTime.current >= minRenderInterval) {
      lastUpdateTime.current = time;
      forceUpdate();
    } else {
      updateTimeout.current = setTimeout(
        () => {
          lastUpdateTime.current += minRenderInterval;
          forceUpdate();
        },
        lastUpdateTime.current + minRenderInterval - time,
      );
    }
  }, [forceUpdate]);

  useEffect(() => {
    objectPanelEventEmitter.on('UPDATE_OBJECT_PANEL', updateObjectPanel);

    return () => {
      objectPanelEventEmitter.removeListener('UPDATE_OBJECT_PANEL', updateObjectPanel);
    };
  }, [updateObjectPanel]);

  return (
    <ObjectPanelContext.Provider
      value={{
        activeKey,
        dimensionValues: dimensionValues.current,
        getDimensionValues,
        polygonSides,
        updateActiveKey: setActiveKey,
        updateDimensionValues,
        updateObjectPanel,
      }}
    >
      {children}
    </ObjectPanelContext.Provider>
  );
};
