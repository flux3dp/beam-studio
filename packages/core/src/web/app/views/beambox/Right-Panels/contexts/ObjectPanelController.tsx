import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import type {
  DimensionKey,
  DimensionKeyBoolean,
  DimensionKeyNumber,
  DimensionValues,
} from '@core/interfaces/ObjectPanel';

const objectPanelEventEmitter = eventEmitterFactory.createEventEmitter('object-panel');

const updateDimensionValues = (newValue: DimensionValues): void => {
  objectPanelEventEmitter.emit('UPDATE_DIMENSION_VALUES', newValue);
};

/* eslint-disable no-redeclare */
function getDimensionValues(): DimensionValues;
function getDimensionValues(key: DimensionKeyNumber): number;
function getDimensionValues(key: DimensionKeyBoolean): boolean;
function getDimensionValues(key?: DimensionKey): any {
  const response = {
    dimensionValues: {},
  };

  objectPanelEventEmitter.emit('GET_DIMENSION_VALUES', response, key);

  return response.dimensionValues;
}
/* eslint-enable no-redeclare */

const minEventInterval = 50;
let updateObjectPanelTimeout: NodeJS.Timeout | undefined = undefined;

export const updateObjectPanel = (): void => {
  clearTimeout(updateObjectPanelTimeout);

  updateObjectPanelTimeout = setTimeout(() => {
    objectPanelEventEmitter.emit('UPDATE_OBJECT_PANEL');
  }, minEventInterval);
};

const updatePolygonSides = (polygonSides: number): void => {
  objectPanelEventEmitter.emit('UPDATE_POLYGON_SIDES', polygonSides);
};

const updateActiveKey = (activeKey: null | string): void => {
  objectPanelEventEmitter.emit('UPDATE_ACTIVE_KEY', activeKey);
};

const getActiveKey = (): null | string => {
  const response = {
    activeKey: undefined as any,
  };

  objectPanelEventEmitter.emit('GET_ACTIVE_KEY', response);

  return response.activeKey as null | string;
};

export default {
  events: objectPanelEventEmitter,
  getActiveKey,
  getDimensionValues,
  updateActiveKey,
  updateDimensionValues,
  updateObjectPanel,
  updatePolygonSides,
};
