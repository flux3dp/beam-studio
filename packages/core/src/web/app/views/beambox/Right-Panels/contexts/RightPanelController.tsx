import eventEmitterFactory from 'helpers/eventEmitterFactory';
import { PanelType } from 'app/constants/right-panel-types';

const rightPanelEventEmitter = eventEmitterFactory.createEventEmitter('right-panel');

const setDisplayLayer = (val: boolean): void => {
  rightPanelEventEmitter.emit('DISPLAY_LAYER', val);
};

const updatePathEditPanel = (): void => {
  rightPanelEventEmitter.emit('UPDATE_PATH_EDIT_PANEL');
};

const setPanelType = (val: PanelType): void => {
  rightPanelEventEmitter.emit('SET_PANEL_TYPE', val);
};

export default {
  setDisplayLayer,
  setPanelType,
  updatePathEditPanel,
};
