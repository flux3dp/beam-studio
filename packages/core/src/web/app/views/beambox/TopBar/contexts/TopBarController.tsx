import BeamboxGlobalInteraction from '@core/app/actions/beambox/beambox-global-interaction';
import CanvasMode from '@core/app/constants/canvasMode';
import eventEmitterFactory from '@core/helpers/eventEmitterFactory';
import type { IDeviceInfo } from '@core/interfaces/IDevice';

const topBarEventEmitter = eventEmitterFactory.createEventEmitter('top-bar');

const setElement = (elem: Element | null): void => {
  BeamboxGlobalInteraction.onObjectBlur();

  if (elem) BeamboxGlobalInteraction.onObjectFocus([elem]);

  topBarEventEmitter.emit('SET_ELEMENT', elem);
};

const updateTitle = (title: string, isCloudFile: boolean): void => {
  topBarEventEmitter.emit('UPDATE_TITLE', title, isCloudFile);
};

const onTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.on('UPDATE_TITLE', handler);
};

const offTitleChange = (handler: (title: string, isCloudFile: boolean) => void): void => {
  topBarEventEmitter.removeListener('UPDATE_TITLE', handler);
};

const setHasUnsavedChange = (hasUnsavedChange: boolean): void => {
  topBarEventEmitter.emit('SET_HAS_UNSAVED_CHANGE', hasUnsavedChange);
};

const getTopBarPreviewMode = (): boolean => {
  const response = { mode: CanvasMode.Draw };

  topBarEventEmitter.emit('GET_CANVAS_MODE', response);

  return response.mode === CanvasMode.Preview;
};

const getSelectedDevice = (): IDeviceInfo | null => {
  const response = { selectedDevice: null };

  topBarEventEmitter.emit('GET_SELECTED_DEVICE', response);

  return response.selectedDevice;
};

const setSelectedDevice = (device: IDeviceInfo): void => {
  topBarEventEmitter.emit('SET_SELECTED_DEVICE', device);
};

export default {
  getSelectedDevice,
  getTopBarPreviewMode,
  offTitleChange,
  onTitleChange,
  setElement,
  setHasUnsavedChange,
  setSelectedDevice,
  updateTitle,
};
