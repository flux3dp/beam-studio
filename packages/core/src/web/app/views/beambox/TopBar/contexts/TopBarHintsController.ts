import eventEmitterFactory from 'helpers/eventEmitterFactory';

const topBarHintsEventEmitter = eventEmitterFactory.createEventEmitter('top-bar-hints');

const setHint = (hintType: string): void => {
  topBarHintsEventEmitter.emit('SET_HINT', hintType);
};

const removeHint = (): void => {
  topBarHintsEventEmitter.emit('REMOVE_HINT');
};

export default {
  setHint,
  removeHint,
};
