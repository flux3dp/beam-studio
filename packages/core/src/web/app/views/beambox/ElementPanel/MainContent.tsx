import type { ReactNode } from 'react';
import React, { useContext, useEffect, useMemo, useRef } from 'react';

import { Button } from 'antd';
import { sprintf } from 'sprintf-js';

import dialogCaller from '@core/app/actions/dialog-caller';
import ThemedButton from '@core/app/components/common/ThemedButton';
import { ContentType, type MainType } from '@core/app/constants/element-panel-constants';
import { ElementPanelContext } from '@core/app/contexts/ElementPanelContext';
import { isMobile } from '@core/helpers/system-helper';
import useI18n from '@core/helpers/useI18n';

import styles from './Content.module.scss';
import BuiltinElement from './Element/BuiltinElement';
import NPElement from './Element/NPElement';
import GridContent from './GridContent';

interface Props {
  types: MainType[];
}

const MainContent = ({ types }: Props): ReactNode => {
  const { contents, contentType, hasLogin, historyIcons, setActiveMainType, setActiveSubType, setSearchKey } =
    useContext(ElementPanelContext);
  const lang = useI18n().beambox.elements_panel;

  const scrollDivRef = useRef<HTMLDivElement>(null);
  const waitFirstAnimation = useRef(isMobile());
  const shodowRef = useRef<HTMLDivElement>(null);
  const handleShadow = () => {
    if (scrollDivRef.current && shodowRef.current && !waitFirstAnimation.current) {
      if (
        // add extra 5px to fix windows browser precision
        scrollDivRef.current.scrollTop + scrollDivRef.current.clientHeight + 5 >=
        scrollDivRef.current.scrollHeight
      ) {
        shodowRef.current.style.display = 'none';
      } else {
        shodowRef.current.style.display = 'block';
      }
    } else {
      if (shodowRef.current) {
        shodowRef.current.style.display = 'none';
      }

      setTimeout(handleShadow, 500);
    }
  };

  useEffect(() => {
    setTimeout(() => {
      // Display shadow after flaoting panel completly opened
      waitFirstAnimation.current = false;
    }, 500);
  }, []);

  useEffect(() => {
    handleShadow();
    // Check again to make sure all icons loaded
    setTimeout(handleShadow, 300);
    // eslint-disable-next-line hooks/exhaustive-deps
  }, [contents]);

  const isSearch = useMemo(() => contentType === ContentType.Search, [contentType]);
  const isSearchEmpty = useMemo(() => {
    if (!isSearch) return false;

    if (!contents[0]) return true;

    if (contents[0].loading || contents[0].fileNames?.length || contents[0].npIcons?.length) return false;

    return true;
  }, [isSearch, contents]);

  return (
    <div className={styles['shadow-container']}>
      <div className={styles['scroll-content']} onScroll={handleShadow} ref={scrollDivRef}>
        {!hasLogin && (
          <ThemedButton
            block
            className={styles.login}
            onClick={() => dialogCaller.showLoginDialog()}
            theme="yellow"
            type="primary"
          >
            {lang.sign_in}
          </ThemedButton>
        )}
        {isSearch && (
          <>
            <ul className={styles.hints}>
              <li>{lang.search_with_english}</li>
              <li>{lang.search_examples} flower, school building.</li>
            </ul>
            {isSearchEmpty && (
              <>
                {contents[0]?.term && (
                  <div className={styles.empty}>{sprintf(lang.search_no_results, contents[0].term)}</div>
                )}
                <div className={styles.bold}>{lang.browse_categories}</div>
                <div className={styles.categories}>
                  {types.map((key) => (
                    <Button
                      key={key}
                      onClick={() => {
                        setSearchKey(undefined);
                        setActiveSubType(undefined);
                        setActiveMainType(key);
                      }}
                      size="small"
                      type="link"
                    >
                      {lang[key]}
                    </Button>
                  ))}
                </div>
              </>
            )}
          </>
        )}
        <div className={styles['icon-list']}>
          {contentType === ContentType.MainType && historyIcons.length > 0 && (
            <>
              <div className={styles.subtitle}>{lang.recently_used}</div>
              {historyIcons.map(({ npIcon, path }) =>
                path ? (
                  <BuiltinElement key={path.fileName} mainType={path.folder} path={path.fileName} />
                ) : hasLogin ? (
                  <NPElement icon={npIcon!} key={npIcon!.id} />
                ) : null,
              )}
            </>
          )}
          {contents.map((content) => (
            <GridContent content={content} key={content.term} />
          ))}
        </div>
      </div>
      <div className={styles.shadow} ref={shodowRef} />
    </div>
  );
};

export default MainContent;
