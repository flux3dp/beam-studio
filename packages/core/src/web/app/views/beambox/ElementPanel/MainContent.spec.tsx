import React from 'react';

import { fireEvent, render } from '@testing-library/react';

import { ElementPanelContext } from '@core/app/contexts/ElementPanelContext';
import { ContentType, MainTypes } from '@core/app/constants/element-panel-constants';

const mockShowLoginDialog = jest.fn();

jest.mock('@core/app/actions/dialog-caller', () => ({
  showLoginDialog: mockShowLoginDialog,
}));

jest.mock('@core/app/contexts/ElementPanelContext', () => ({ ElementPanelContext: React.createContext({}) }));

jest.mock('./Element/BuiltinElement', () => 'builtin-element');
jest.mock('./Element/NPElement', () => 'NP-element');
jest.mock('./GridContent', () => 'grid-content');

const mockSetActiveMainType = jest.fn();
const mockSetActiveSubType = jest.fn();
const mockSetSearchKey = jest.fn();

import MainContent from './MainContent';

describe('test MainContent', () => {
  it('should render correctly', () => {
    const { container } = render(
      <ElementPanelContext.Provider
        value={{ contents: [{}, {}], contentType: ContentType.MainType, hasLogin: true, historyIcons: [] } as any}
      >
        <MainContent types={[]} />
      </ElementPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render correctly when not login', () => {
    const { container } = render(
      <ElementPanelContext.Provider
        value={{ contents: [{}, {}], contentType: ContentType.MainType, hasLogin: false, historyIcons: [] } as any}
      >
        <MainContent types={[]} />
      </ElementPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();

    const loginButton = container.querySelector('.login');

    fireEvent.click(loginButton!);
    expect(mockShowLoginDialog).toHaveBeenCalledTimes(1);
  });

  it('should render history correctly', () => {
    const { container } = render(
      <ElementPanelContext.Provider
        value={
          {
            contents: [{}, {}],
            contentType: ContentType.MainType,
            hasLogin: true,
            historyIcons: [
              { path: { fileName: 'icon-circle', folder: 'basic' }, type: 'builtin' },
              { npIcon: { id: '1234', thumbnail_url: 'url_1234' }, type: 'np' },
              { path: { fileName: 'i_circular-1', folder: 'decor' }, type: 'builtin' },
              { npIcon: { id: '4321', thumbnail_url: 'url_4321' }, type: 'np' },
            ],
          } as any
        }
      >
        <MainContent types={[]} />
      </ElementPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();
  });

  it('should render search hint correctly', () => {
    const { container } = render(
      <ElementPanelContext.Provider
        value={
          {
            contents: [],
            contentType: ContentType.Search,
            hasLogin: true,
            historyIcons: [],
            setActiveMainType: mockSetActiveMainType,
            setActiveSubType: mockSetActiveSubType,
            setSearchKey: mockSetSearchKey,
          } as any
        }
      >
        <MainContent types={MainTypes} />
      </ElementPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();

    const typeButton = container.querySelectorAll('.categories > button')[3];

    fireEvent.click(typeButton!);
    expect(mockSetActiveMainType).toHaveBeenCalledWith(MainTypes[3]);
    expect(mockSetActiveSubType).toHaveBeenCalledWith(undefined);
    expect(mockSetActiveSubType).toHaveBeenCalledWith(undefined);
  });

  it('should render search hint correctly when no result', () => {
    const { container } = render(
      <ElementPanelContext.Provider
        value={
          {
            contents: [{ term: 'search' }],
            contentType: ContentType.Search,
            hasLogin: true,
            historyIcons: [],
            setActiveMainType: mockSetActiveMainType,
            setActiveSubType: mockSetActiveSubType,
            setSearchKey: mockSetSearchKey,
          } as any
        }
      >
        <MainContent types={MainTypes} />
      </ElementPanelContext.Provider>,
    );

    expect(container).toMatchSnapshot();

    const typeButton = container.querySelectorAll('.categories > button')[3];

    fireEvent.click(typeButton!);
    expect(mockSetActiveMainType).toHaveBeenCalledWith(MainTypes[3]);
    expect(mockSetActiveSubType).toHaveBeenCalledWith(undefined);
    expect(mockSetActiveSubType).toHaveBeenCalledWith(undefined);
  });
});
