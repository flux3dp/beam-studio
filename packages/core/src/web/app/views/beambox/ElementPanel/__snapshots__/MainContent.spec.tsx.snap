// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test MainContent should render correctly 1`] = `
<div>
  <div
    class="shadow-container"
  >
    <div
      class="scroll-content"
    >
      <div
        class="icon-list"
      >
        <grid-content
          content="[object Object]"
        />
        <grid-content
          content="[object Object]"
        />
      </div>
    </div>
    <div
      class="shadow"
      style="display: none;"
    />
  </div>
</div>
`;

exports[`test MainContent should render correctly when not login 1`] = `
<div>
  <div
    class="shadow-container"
  >
    <div
      class="scroll-content"
    >
      <button
        class="ant-btn css-dev-only-do-not-override-hash ant-btn-primary ant-btn-color-primary ant-btn-variant-solid ant-btn-block login yellow"
        type="button"
      >
        <span>
          Sign in to get 10,000+ free icons
        </span>
      </button>
      <div
        class="icon-list"
      >
        <grid-content
          content="[object Object]"
        />
        <grid-content
          content="[object Object]"
        />
      </div>
    </div>
    <div
      class="shadow"
      style="display: none;"
    />
  </div>
</div>
`;

exports[`test MainContent should render history correctly 1`] = `
<div>
  <div
    class="shadow-container"
  >
    <div
      class="scroll-content"
    >
      <div
        class="icon-list"
      >
        <div
          class="subtitle"
        >
          Recently Used
        </div>
        <builtin-element
          maintype="basic"
          path="icon-circle"
        />
        <np-element
          icon="[object Object]"
        />
        <builtin-element
          maintype="decor"
          path="i_circular-1"
        />
        <np-element
          icon="[object Object]"
        />
        <grid-content
          content="[object Object]"
        />
        <grid-content
          content="[object Object]"
        />
      </div>
    </div>
    <div
      class="shadow"
      style="display: none;"
    />
  </div>
</div>
`;

exports[`test MainContent should render search hint correctly 1`] = `
<div>
  <div
    class="shadow-container"
  >
    <div
      class="scroll-content"
    >
      <ul
        class="hints"
      >
        <li>
          Please type your search in English.
        </li>
        <li>
          Examples:
           flower, school building.
        </li>
      </ul>
      <div
        class="bold"
      >
        Browse Categories
      </div>
      <div
        class="categories"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Basic
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Decor
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Animal
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Holiday
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Nature
          </span>
        </button>
      </div>
      <div
        class="icon-list"
      />
    </div>
    <div
      class="shadow"
      style="display: none;"
    />
  </div>
</div>
`;

exports[`test MainContent should render search hint correctly when no result 1`] = `
<div>
  <div
    class="shadow-container"
  >
    <div
      class="scroll-content"
    >
      <ul
        class="hints"
      >
        <li>
          Please type your search in English.
        </li>
        <li>
          Examples:
           flower, school building.
        </li>
      </ul>
      <div
        class="empty"
      >
        No results were found for "search". Please check spelling or try a different keyword.
      </div>
      <div
        class="bold"
      >
        Browse Categories
      </div>
      <div
        class="categories"
      >
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Basic
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Decor
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Animal
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Holiday
          </span>
        </button>
        <button
          class="ant-btn css-dev-only-do-not-override-hash ant-btn-link ant-btn-color-primary ant-btn-variant-link ant-btn-sm"
          type="button"
        >
          <span>
            Nature
          </span>
        </button>
      </div>
      <div
        class="icon-list"
      >
        <grid-content
          content="[object Object]"
        />
      </div>
    </div>
    <div
      class="shadow"
      style="display: none;"
    />
  </div>
</div>
`;
